File: mobile/ios/OmenIOS/OmenIOS/App/UpdateGateViewModel.swift
import Foundation

enum UpdateGateState: Equatable {
    /// Includes both "checking" and "check failed/unavailable" — a network error must never
    /// block the app, so both resolve to the same pass-through state as a successful "ok".
    case passed
    case blocked(minimumVersion: String)
}

@MainActor
final class UpdateGateViewModel: ObservableObject {
    @Published private(set) var state: UpdateGateState = .passed

    private let client: MinVersionGateChecking
    private let currentVersion: String
    private let platform = "ios"

    init(client: MinVersionGateChecking, currentVersion: String = AppVersionProvider.current) {
        self.client = client
        self.currentVersion = currentVersion
    }

    func check() async {
        switch await client.check(platform: platform, currentVersion: currentVersion) {
        case .ok, .unavailable:
            state = .passed
        case .updateRequired(let minimumVersion):
            state = .blocked(minimumVersion: minimumVersion)
        }
    }
}

File: mobile/android/app/src/main/kotlin/com/slopssaloon/omen/app/feature/api/UpdateGateViewModel.kt
package com.slopssaloon.omen.app.feature.api

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/**
 * O7 — forced-update gate state. iOS mirror: `App/UpdateGateViewModel.swift`.
 */
sealed interface UpdateGateState {
    /**
     * Covers "not yet checked", "check succeeded and we're supported", and "check failed".
     * A failed check must never block, so it collapses into the same pass-through state.
     */
    data object Passed : UpdateGateState
    data class Blocked(val minimumVersion: String) : UpdateGateState
}

class UpdateGateViewModel(
    private val client: MinVersionGateChecking,
    private val currentVersion: String,
) {
    var state: UpdateGateState by mutableStateOf(UpdateGateState.Passed)
        private set

    suspend fun check() {
        state = when (val result = client.check(platform = "android", currentVersion = currentVersion)) {
            is MinVersionGateResult.UpdateRequired -> UpdateGateState.Blocked(result.minimumVersion)
            MinVersionGateResult.Ok, MinVersionGateResult.Unavailable -> UpdateGateState.Passed
        }
    }
}
